### ISSUE TITLE
Endless spinner on OC page, for Producer with "Profile only" package

### ISSUE THREAD

**Issue by filipefurtad0** [openfoodfoundation/openfoodnetwork#5831]
## Description

Producers which choose the "Profile only" package, don't have access to an OC page, by default. This option is not visible on their dashboard.

If a producer attributes the necessary permissions for a hub to sell its produce, then the Order Cycle option will become visible in the Dashboard. After this, visiting the OC page will reveal the order cycles in which the producer was added. However, this page now displays an endless spinner (as if has not yet fully loaded).

Role permissions which appear correct:
- the producer sees the OCs he was added into, but cannot change dates
- the producer can edit each OC, and individually select which products to include

## Expected Behavior

The OC page should not display an endless spinner, for producers with "Profile only" package, included in a given OC.

## Actual Behaviour

The OC page should not display an endless spinner.

## Steps to Reproduce

1. Create or use a Producer with "Profile only" package. Check that the dashboard does not display the option to visit the page "Order Cycles"
2. As a producer: Add permissions for a given Hub to add this Producer to a given OC.
3. As a manager of this hub, add the producer to an OC

As the producer:
4. It should now be possible now to navigate to the OC page.
5. See the endless spinner on the page /admin/order_cycles

## Animated Gif/Screenshot

![image.png](https://images.zenhubusercontent.com/5d761b049dde650001eaf32f/173d0317-eb4b-4080-9c6e-a4f355c4e6d2)

## Workaround

No workaround, but no serious impact

## Severity

bug-s5: we can live with it, only a few users impacted

## Your Environment

*    Version used: v3.1.3
*    Browser name and version: Firefox 78
*    Operating System and version (desktop or mobile): Desktop Ubuntu 20.04

**Comment by RachL** [comment]
@filipefurtad0 Aah interesting! But I wonder if this is a bug then. I don't think this use case exist? With profile only you cannot create product, correct?

**Comment by filipefurtad0** [comment]
Hey @RachL ,
I used this profile for both this issue and #5832:

- All Rooty Producer <my_email>+allrootyp@g....., in staging-uk

This account is set to be able to produce, without having its own a shopfront (Profile Only); but it is able to create products.

**Comment by RachL** [comment]
@filipefurtad0 I'm unable to reproduce when the producer has product. I will try to test when there is no product and let you know :+1:

**Comment by filipefurtad0** [comment]
It is observable here as well, on staging-FR, different hub/producer combination (at least on my system):

![Peek 2020-07-27 23-01](https://user-images.githubusercontent.com/49817236/88596530-2c2e7700-d05d-11ea-9d26-1202fc39a0e2.gif)

**Comment by RachL** [comment]
@filipefurtad0 now I understand, can you try and switch to profile that sells? I believe it will fix the problem. Profile only can reate product but they cannot sell them, this is why OC page is acting weirdly.

**Comment by filipefurtad0** [comment]
Hey @RachL ,

Yes, that's it: switching to a producer shop profile keeps all rights access to order cycles created/managed by another hub (producer can still see, edit the OC, tick/untick his products from incoming/outgoing) but the endless spinner disappears.

So the bug only seems to happen for producers which cannot sell his own products but have their products added to an OC.

**Comment by RachL** [comment]
> So the bug only seems to happen for producers which cannot sell his own products but have their products added to an OC.

Yes so I'm happy we keep it as s5 but I think there is really little chance that we will fix it. Maybe I'm wrong but profile only are designed so they cannot sell products, even through a hub. If we enable profile only to sell, there won't be a difference with producer package anymore.

I guess that the fix should be that a profile only cannot access enterprise_relationship page to avoid this?

**Comment by filipefurtad0** [comment]
Interesting!
I think the case you describe would be for a "non-producer - profile only". In that case, I agree that this type of user should not have access to the enterprise_relationship page - which apparently it does, so this would be a separate issue. I'll double check and open one!

However, I understood a "producer - profile only" as the role meant to farmers/producers focused on production of goods only, while allowing to sell his produce but not independently. It cannot have it's own shopfront, so they rely on other enterprises to promote and sell its produce. This is my interpretation of this section of the user guide:
https://guide.openfoodnetwork.org/your-quick-start-on-ofn-given-who-you-are#profile

So, a "fresh" producer only cannot see "Order Cycles" in the main menu, or create one; He only sees this after he granted permissions an was added.

A producer which is able to sell (producer - shop) not only produces, but is able to create order cycles himself, and have his own shopfront.

What do you think?

**Comment by RachL** [comment]
@filipefurtad0 hmmm ok interesting, I didn't know that. I guess there is something weird if the producer does it himself, which is not a case we have often. I think in France 99% of producer profile only are created by hubs directly, and in that case I can't reproduce the issue if I grand access to the producer. So there is maybe a step missing if the producer does the profile only himself. I will test a bit deeper thanks :+1:

**Comment by abdellani** [comment]
Hi @RachL  @filipefurtad0 

The problem in summary:
The `RequestMonitor` has a counter to track the active API calls.
if any query fails, the counter will not be decremented.

For example: 
if you have 3 API calls. The initial counter will start with 3.
if one of the calls fails. The counter will only be decremented 2 times (final value will be 1).

The long version: 
1. The spinner component use `RequestMonitor` (app/assets/javascripts/admin/index_utils/services/request_monitor.js.coffee) to decide whether it should show the spinner or not.
2. `RequestMonitor` has a counter to track the running API calls.
3. When the page is loaded, The `OrderCyclesCtrl` run multiple API calls (which are tracked by  `RequestMonitor`)
4. One of those API calls target: `http://localhost:3000/admin/enterprises/visible.json?ams_prefix=basic`
5. This one will fail due to a permission issue. (producer can't access the enterprise list)
6. The spinner will never be removed because of the `RequestMonitor` counter.



I guess this issue affects all the other pages that rely on `RequestMonitor` to render the spinner.
